comprehension/03_multiple_methods.md

```diff
@@ -0,0 +1,238 @@
+1. After you feel that you understand the code (or are as close as you can get), look at the provided list of input values. For each value, try to work out what the return value of the method will be for that input. **Write down your expected return values.**
+1. Finally, copy the code snippet into IRB and then try calling the method with each input:
+    1. Was your expectation of the return value correct?
+    1. If not, try to look at the code again and figure out what it's doing differently from how you originally understood it.
+    1. If you figure out something knew about the code (i.e. you understand it better now), rewrite your expanation of the code from the first step. Also update any of the expected return values if you now believe they will be different.
+
+### Tips
+1. While you should push yourself to predict the correct return value for each of these problems, this is intentionally a challenging exercise. Don't be discouraged if you get something wrong! The instructions above involve re-reading and comprehending the code again after trying each input precisely for this reason. Look at this as an iterative process, each time hopefully getting closer to a full understanding of the code snippet.
+1. You CAN use the Internet, any of our lecture resources, or any of your notes. Some of these problems might involve methods that you've not seen in class, so you will need to look them up in the Ruby docs.
+1. Some of the input values for a given problem might result in an error rather than a return value. This is intentional.
+1. Each problem is intended to be more challenging to read and comprehend than the previous ones.
+
+## Practice Problems
+Use the following data for all problems in this worksheet:
+
+```ruby
+VEHICLE_INFO = {
+  coupe: {
+    seats: 2,
+    fwd: false,
+    cargo: :small,
+  },
+  sedan: {
+    seats: 4,
+    fwd: false,
+    cargo: :small,
+  },
+  minivan: {
+    seats: 6,
+    fwd: false,
+    cargo: :medium,
+  },
+  truck: {
+    seats: 2,
+    fwd: true,
+    cargo: :large,
+  },
+}
+
+CARGO_SIZES = [:none, :small, :medium, :large, :xlarge]
+
+VEHICLE_DB = {
+  alamo: {
+    coupe: 5, sedan:  6, minivan: 4, truck: 0
+  },
+  budget: {
+    coupe: 0, sedan: 14, minivan: 1, truck: 1
+  },
+  enterprise: {
+    coupe: 6, sedan:  8, minivan: 0, truck: 9
+  },
+  hertz: {
+    coupe: 7, sedan: 10, minivan: 2, truck: 4
+  },
+}
+```
+
+### Problem 1
+#### Code Snippet
+```ruby
+def enough_seats?(type, seats_required)
+  VEHICLE_INFO[type][:seats] >= seats_required
+end
+
+def has_fwd?(type)
+  VEHICLE_INFO[type][:fwd]
+end
+
+def matches_search?(type, seats_req, fwd_req)
+  enough_seats?(type, seats_req) && (!fwd_req || has_fwd?(type))
+end
+
+def vehicles_available(seats_required, fwd_required=false)
+  VEHICLE_DB.values.map do |inventory|
+    inventory.select do |type, _|
+      matches_search?(type, seats_required, fwd_required)
+    end.values
```

Comment:
it might be better to just pull these apart?

Reply:
Do you mean specifically the chained together `end.values` bit or something else?

Reply:
yes! the `end.values` part